### ISSUE TITLE
RFC better local reproducible tests

### ISSUE THREAD

**Issue by david-caro** [inveniosoftware/cookiecutter-invenio-module#66]
Use tools like tox and docker-compose to achieve maximum test run reproducible results, to allow locally testing and debugging the issues.

This is implemented on pr #65

**Comment by jirikuncar** [comment]
Note that we used to have both tox and docker-compose in individual packages in the past. I'm in favor in returning tox as long as we don't duplicate testing logic.

The **only** thing one has to run in Invenio software repositories is `./run-tests.sh`.

--
cc @inveniosoftware/triagers

**Comment by david-caro** [comment]
sohuld that be using the reproducible docker-compose or the lighter tox? Maybe allow passing an option for it?

**Comment by lnielsen** [comment]
As @jirikuncar said we originally had both. However, they were rarely used and thus weren't always working. I think they weren't used because they are simply too slow. Personally I very rarely run the entire test matrix locally. I prefer simply pushing a branch to github and let Travis do the job. 

In case Travis report a problem for one of the builds that doesn't match my local install, I usually find it pretty fast to create a virtualenv and install the package. Also, you usually only need to run one or two particular tests over and over again, until the problem is solved.

So long story short, IMHO I don't think it's worthwhile the effort to add these.

**Comment by david-caro** [comment]
This is very interesting for integration tests too, where you have multiple agents (postgres, elasticsearch, celery, rabbitmq, worker node, ...). Setting up all that environment in your own laptop it's hard, and having to maintain/sync it every time way harder.

**Comment by jirikuncar** [comment]
> This is very interesting for integration tests too, ...

Sure. We are providing Docker recipe for overlays.

> I think they weren't used because they are simply too slow. 

👍

**Comment by david-caro** [comment]
Running the invenio-workflows basic (pytest on sqlite) recreating the venv/docker container once the deps are cached:
* on tox:   0m26.087s
* plain venv: 0m31.963s
* docker: 1m6.445s (it does not actually use the pip cache, yet at least)

Once you have run them once, you can enter the virtualenv of any of them:

* tox: source .tox/$suite/bin/activate
* plain venv: workon $venvname or source path/to/venv/bin/activa
* docker: docker-compose -f docker-compose.test.yaml run --rm test-$suite bash, and once there, source .tox/$suite/bin/activate

And once there, running the tests on any of the envs is technically the exact same process, so the same exact speed (just run py.test... or ./run_tests.sh or whatever you want to run there).

**Comment by david-caro** [comment]
And on travis, the full suite takes ~3min fully recreating the environment with docker (including the databases) and running tox on it:

   https://travis-ci.org/inveniosoftware/invenio-workflows/builds/149122031

**Comment by david-caro** [comment]
Where a previous run without tox nor docker took >4min:

   https://travis-ci.org/inveniosoftware/invenio-workflows/builds/76413761
